File: examples/go/p2pkh/P2PKH.tsop.go
package contract

import "tsop"

type P2PKH struct {
	tsop.SmartContract
	PubKeyHash tsop.Addr `tsop:"readonly"`
}

func (c *P2PKH) Unlock(sig tsop.Sig, pubKey tsop.PubKey) {
	tsop.Assert(tsop.Hash160(pubKey) == c.PubKeyHash)
	tsop.Assert(tsop.CheckSig(sig, pubKey))
}

File: examples/rust/p2pkh/P2PKH.tsop.rs
use tsop::prelude::*;

#[tsop::contract]
pub struct P2PKH {
    #[readonly]
    pub pub_key_hash: Addr,
}

#[tsop::methods(P2PKH)]
impl P2PKH {
    #[public]
    pub fn unlock(&self, sig: &Sig, pub_key: &PubKey) {
        assert!(hash160(pub_key) == self.pub_key_hash);
        assert!(check_sig(sig, pub_key));
    }
}
